### PR TITLE
feat: Add trailing comma in multiline to PER-CS 2.0

### LINKED ISSUE #7664: per-cs: add trailing_comma_in_multiline rule
`trailing_comma_in_multiline` (fully or configured) can become part of PER-CS
https://www.php-fig.org/per/coding-style/#26-trailing-commas

### PR DESCRIPTION
**Pull request by michaelvickersuk**
As per https://www.php-fig.org/per/coding-style/#26-trailing-commas

> Numerous PHP constructs allow a sequence of values to be separated by a comma, and the final item may have an optional comma ... If the list is split across multiple lines, then the last item MUST have a trailing comma.

Closes #7664

### CODE REVIEW

**Comment by michaelvickersuk** [comment]
I'll need to take some advice on how's best to pass the PHP 7.4 tests, will it be simply a case of removing the `parameters` and `match` elements or can we somehow still include them when PHP < 8.0?

File: tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php
Comment (julienfalque) [inline tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php:76]: This will fail on PHP 7.4 because trailing comma is not allowed here. You need to duplicate this test so that one runs on PHP 7.4 (no comma added here) and the other on PHP 8.0+ (comma added). You can have a look at `tests/Fixtures/Integration/set/@PHP74Migration.test` for an example.

File: src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php
Comment (keradus) [reply src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php:111]: please elaborate on reasoning.

did you had a chance to compare with other rules with similar concern?
I would love to have overview and align on consensus for all rules (perhaps out of scope of this PR)

File: src/RuleSet/Sets/PERCS2x0Set.php
Comment (keradus) [reply src/RuleSet/Sets/PERCS2x0Set.php:48]: > PER rule covers any kind of multiline list

I'm not sure. eg, it mentions function arguments in function declaration, but not function call params.

I would be careful with `@all` by default, as it can easily create bc breaks in someone CS.  I'm +1 to have `@all` available

File: tests/Fixer/ControlStructure/TrailingCommaInMultilineFixerTest.php
Comment (keradus) [inline tests/Fixer/ControlStructure/TrailingCommaInMultilineFixerTest.php:32]: there should be a test for <8 (ie to ensure we do not produce invalid code under <8 runtime)

File: tests/Fixtures/Integration/set/@PER-CS2.0.test
Comment (keradus) [inline tests/Fixtures/Integration/set/@PER-CS2.0.test:6]: nope.
if you need to test specific part of ruleset that requires specific engine, test that part of ruleset under specific engine. 

now, you said that the ruleset is not tested <8 at all.

eg ` @PSR12_php71.test` test, as example

Comment (julienfalque) [reply tests/Fixtures/Integration/set/@PER-CS2.0.test:6]: I'm not sure I understand. There also is a ruleset test for PHP 7.4 specifically.

File: src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php
Comment (julienfalque) [reply src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php:111]: I think it might be better for end users to always allow the `parameters` config and simply ignore it on PHP 7.4. But yeah that will require to reconsider our approach globally. Let's keep current rule level behavior. The rule should be configured automatically at ruleset level.

**Review by julienfalque — APPROVED** [review]
LGTM 👍

@keradus Do you want to check again before merging?

**Review by Wirone — APPROVED** [review]

File: tests/Fixtures/Integration/set/@PER-CS2.0.test
Comment (keradus) [reply tests/Fixtures/Integration/set/@PER-CS2.0.test]: we shall test as much as possible on lowest PHP version, and on high PHP version check only the "new part".

instead, the PR test as much as possible on "high PHP", and on low PHP the "old part".

File: doc/ruleSets/PER-CS2.0.rst
Comment (keradus) [inline doc/ruleSets/PER-CS2.0.rst:25]: this is super confusing for me. ie, on PHP <8, the rule is not configured with `parameters`

File: tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php
Comment (keradus) [inline tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php:72]: I'm not sure about `arguments` config belonging to PER-CS2.
It's also not listed as example in https://www.php-fig.org/per/coding-style/#26-trailing-commas .
Can you raise a PR to above examples to include the function call arguments ?

I'm all in for that style personally, but I'm not sure it's indeed part of PER-CS. Same time, i know big communities against it (eg Symfony)

File: src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php
Comment (keradus) [reply src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php]: > reconsider our approach globally

i think it may be interesting part to go with. Would you mind to create follow-up ticket to have that investigation?

File: tests/Fixer/ControlStructure/TrailingCommaInMultilineFixerTest.php
Comment (keradus) [reply tests/Fixer/ControlStructure/TrailingCommaInMultilineFixerTest.php]: solved

File: tests/Fixtures/Integration/set/@PER-CS2.0.test
Comment (keradus) [reply tests/Fixtures/Integration/set/@PER-CS2.0.test]: with new commits, I see it solved 👍🏻

**Comment by keradus** [comment]
> Do you want to check again before merging?

I don't mind, but as you poked me it got my attention anyway.

old concerns of mine are solved 👍🏻 
i found 2 new.

File: tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php
Comment (keradus) [reply tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php:72]: (originally concern raised in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7916#discussion_r1546759323 )

File: doc/ruleSets/PER-CS2.0.rst
Comment (julienfalque) [reply doc/ruleSets/PER-CS2.0.rst:25]: The fact that some elements are not enabled on PHP 7.4 is implicit and could be seen as an implementation detail of the ruleset. I agree it would be better to make it explicit in the doc somehow, but to me it's not a requirement to merge this PR.

File: tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php
Comment (julienfalque) [reply tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php:72]: > It's also not listed as example in https://www.php-fig.org/per/coding-style/#26-trailing-commas .

It's listed in the first paragraph though:
> Examples include array key/value pairs, **function arguments**, closure use statements, match() statement branches, etc.

Again, IMO the following covers any existing and future list that allows trailing comma after last item:

> Numerous PHP constructs allow a sequence of values to be separated by a comma, and the final item may have an optional comma.
> […]
> If the list is split across multiple lines, then the last item MUST have a trailing comma.

So I would include function arguments even if not listed explicitly.

File: src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php
Comment (julienfalque) [reply src/Fixer/ControlStructure/TrailingCommaInMultilineFixer.php]: Sure: #7919.

File: tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php
Comment (keradus) [reply tests/Fixtures/Integration/set/@PER-CS2.0.test-out.php:72]: We had lengthy discussion on `function arguments` on Sf repo - does it apply to function definition `function foo($arg) {...}` , call `foo(1)` or both.

I recognise that some communities are against that rule and claiming that PER is not conclusive on it - thus idea to enrich the examples in PER docs ;)

but anyway, as I said I'm happy for that style myself, I'm happy we do not enable it for Sf ruleset.

if we enable it by default to PER ruleset - in case ppl would complain (i hope not), let's drop that one config item from ruleset and push to extend PER docs (and bring it back to ruleset)

**Review by keradus — APPROVED** [review]

**Comment by julienfalque** [comment]
Thanks @michaelvickersuk!

**Comment by raveren** [comment]
This is an another (see #6933) BC break in a minor version.

My config:

```php
return (new PhpCsFixer\Config())->setRules([
    '@PER-CS'                           => true,
    'not_operator_with_successor_space' => true,
    'no_break_comment'                  => false,
])
    ->setFinder($finder)
    ->setCacheFile(__DIR__ . '/storage/.php-cs-fixer.cache')
;
```

Now I have 5% of my codebase "breaking" thanks to this PR.

**Comment by kubawerlos** [comment]
> This is an another (see #6933) BC break in a minor version.

@raveren what the h. do you mean? How adding new fixer is BC break?

**Comment by raveren** [comment]
What I mean is, people set up CI pipelines to analyze and deploy their code. (P.S. why else do people use php-cs-fixer for?)

Someone in the team does an innocent composer update and CI pipeline is broken because 5% of the codebase is now not passing the cs-fixer step.

I meant to reference issue #6658 in my original comment, it's the second time in less than two years that this scenario happens.

**Comment by kubawerlos** [comment]
> Someone in the team does an innocent composer update and CI pipeline is broken

Sound to me like the problem is not on PHP CS Fixer side 😉. Maybe you need to have constraint like `~3.52.0`.

Also, similarly to #6658, the same reasoning @keradus made there, according to https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/feature-or-bug.rst#feature this is no BC break, but more of a feature, and I agree with it.

You are using PER and E stands for evolving, and in this case it simply evolved.

**Comment by Wirone** [comment]
@raveren You need to decide whether you follow the `PER-CS` standard or not. In fact, there is #4502 and #7247 that, when implemented, will additionally change the scope of that rulesets. You have 3 options:

- lock Fixer version to keep changes under control
- accept that these rulesets are evolving and may be modified according to the specs that are _not_ covered fully now
- prepare your own ruleset, with 100% explicit config of rules with options (rules are under BC promise, rulesets not)

Thank you for understanding and in the future, please consider interacting without accusations.

**Comment by raveren** [comment]
> Maybe you need to have constraint like ~3.52.0.

The readme states this project is using SemVer 
image

so it's wiser to take in nice feedback from your patient users, accumulate that knowledge and hopefully it will someday fruit into a professional BC policy for the project

**Comment by raveren** [comment]
replying to @Wirone in that case maybe it would be nice then to notify your users on the website (https://cs.symfony.com/doc/ruleSets/PER-CS.html) that the rulesets can and will change under their noses, and thus are not advised to be used in CI pipelines?

**Comment by raveren** [comment]
I think this info is hugely important:

> rules are under BC promise, rulesets not

I can't find that stated anywhere, and that already constitutes a constructive and valid BC policy

**Comment by Wirone** [comment]
> replying to @Wirone in that case maybe it would be nice to notify your users on the website ([cs.symfony.com/doc/ruleSets/PER-CS.html](https://cs.symfony.com/doc/ruleSets/PER-CS.html)) that the rulesets can and will change under their noses, and thus are not advised to be used in CI pipelines?

Why would it be not advised? If you follow pre-defined ruleset, you agree to its rules. It's not a problem of the tool, it's the problem of your CI pipeline and development process. There is nothing like "someone in the team does an innocent composer update" - if there was Composer update and Fixer's version was bumped, developer who did the update should apply CS rules, so nothing fails in CI. If applied changes are not wanted, then your own config should be improved to continue with what you had before (override rulesets' config with custom config for specific rules).

> I can't find that stated anywhere, and that already constitutes a constructive and valid BC policy.
> But users need to know that to prevent such unpleasant surprises in the future.

I don't remember now if we have such explicit information about rulesets, but I always thought it's obvious. Imagine bumping major version after every ruleset modification, or adding more and more clones of existing rulesets 🤷. You want stable ruleset? Prepare your own.

**Comment by raveren** [comment]
>  I always thought it's obvious.
well I can vouch to you that it is not :)
image

P.S. It's made especially difficult since php-cs-fixer can't be made compatible with phpstorm's code formatter, and to fix each new rule by cs-fixer I need to come up with new and inventive ways to format edge cases that work for both :) espacially in dust filled legacy corners of the application.

Here's an example: the only well looking formatting for this case accepted by both:

```php
                $line =
                    [
                        'key'     => $value,
                        'key2'    => $value2,
                    ]
                    + $line;
```

Not your fault, just giving feedback, on what real world is like for your users.

**Comment by Wirone** [comment]
> P.S. It's made especially difficult since php-cs-fixer can't be made compatible with phpstorm's code formatter, and to fix each new rule by cs-fixer I need to come up with new and inventive ways to format edge cases that work for both :) 

It's not true. You can set up external formatter and apply Fixer changes in real-time. I use it for developing Fixer:

image

It either applies the rules automatically, or underlines code that violates the standard and provides controls to apply fixes:

image

It's possible to make it consistent with CI and fully automated (with actions executed on commit).

**Comment by kubawerlos** [comment]
> import cleanup

can you elaborate how import cleanup is not supported?

> max line length

that's by design, because it can never be fully supported

**Comment by raveren** [comment]
Import cleanup is sorting and removing the unused `use` statements, it's a minor thing, but then again each individual rule is.

**Comment by Wirone** [comment]
> Import cleanup is sorting and removing the unused `use` statements, it's a minor thing, but then again each individual rule is.

You mean [this](https://cs.symfony.com/doc/rules/import/no_unused_imports.html)? 🤷‍♂️

> Plus: the inbuilt formatter is simpler to set up and use

How's sharing parts of `.idea` config and/or formatter config that can be imported to IDE easier than shared config for CS Fixer? I've been using Fixer or ECS for years now, automated locally and in central CI and never had problems with conflicts between PHPStorm's built-in formatter and the external tooling. I can't be sure, but maybe you have invalid setup that leads to these kind of problems?

> which is especially relevant when we're talking about a rotating team of developers on different platforms: some might not even have PHP running locally.

Set up Docker stack and configure remote PHP interpreter within shared `.idea` config, so everyone can run everything on the same runtime - problem solved. Anyway, what's the point in working with the PHP code and doing changes, when you don't have PHP runtime and can't run CS fixes and tests?

**Comment by Wirone** [comment]
> Besides, no hard feelings, but the number one reason I wouldn't base so much of a business on this project is because it broke its SemVer (and my trust) twice and the support response both times was "no we didn't! 👺".

Modifying built-in rulesets is not a SemVer violation, as the engine (public API) does not change and each specific rule also works the same. The only thing that changes is the scope of the ruleset, which you implicitly agree on by opting-in when you base your setup on such ruleset. It's like complaining that PHPStan reports new issues when you have bleeding edge enabled. Sorry if it was not clear, we'll try to add such information in the docs. However, stating that Fixer violates SemVer after all the effort we put into keeping backward compatibility is just rude.

**Comment by kubawerlos** [comment]
@raveren thank you very much for this conversation.

It is a nice feeling - knowing that this library helps hundreds or thousands of users, who are not complaining about what we do here (do I have to remind you that we do it in our spare time, and for free?), and one complaining ignorant (yes, you are, you complained about missing features that are present here for years) once every few months.

Thank you once again.

P.S. I'm adding:

> Someone in the team does an innocent composer update and CI pipeline is broken

to the list of my favourite IT quotes.

**Comment by julienfalque** [comment]
> just giving feedback, on what real world is like for your users

I mean, yeah, none of the maintainers here has a "real world" job where they could use the thing they build and see how it actually works 🤷

Also, even a little bugfix can break a CI. This is true for any package you might rely on but especially for us. Like e.g. a bug that prevents a rule from fixing code in an edge-case that happens to exist in your codebase. Fixing that bug will make your CI fail too, this does't mean we should make a major release each time we fix a bug. So please, pin the versions of your tools, check what happens when you do upgrade them, and be nice when reporting actual issues. Thanks.

**Pull request merged**
